File: backend/services/business/reconciliation/pos-closing-credit-validation/app/domain/e_ticket.py
"""O e-Ticket de um caso — a referência do pedido aberto na SIMO."""

import re

from app.domain.errors import InvalidETicketError

MAX_E_TICKET_LENGTH = 40

_E_TICKET = re.compile(r"[A-Za-z0-9][A-Za-z0-9._/-]*")


def normalize_e_ticket(raw: object) -> str | None:
    """O e-Ticket pronto a guardar, ou `None` para o apagar."""
    if raw is None:
        return None
    if not isinstance(raw, str):
        raise InvalidETicketError("O e-Ticket tem de ser texto.")

    value = raw.strip()
    if not value:
        return None
    if len(value) > MAX_E_TICKET_LENGTH:
        raise InvalidETicketError(f"O e-Ticket tem no máximo {MAX_E_TICKET_LENGTH} caracteres.")
    if not _E_TICKET.fullmatch(value):
        raise InvalidETicketError(
            "O e-Ticket só pode ter letras, algarismos e os separadores - _ / . "
            "— e tem de começar por letra ou algarismo."
        )
    return value

File: frontend/src/app/features/payments-and-channels/channels/pos/pos-closing-credit-validation/data/e-ticket.ts
/**
 * A forma de um e-Ticket — espelha o `domain/e_ticket.py` do serviço, como o
 * `sla.ts` espelha o `sla.py`.
 *
 * **Quem protege é o servidor.** Isto só serve para o operador saber o que está
 * mal enquanto escreve, em vez de carregar em «Guardar» e receber a recusa. As
 * regras têm de ser as mesmas dos dois lados; se mudarem, mudam lá primeiro.
 */

export const MAX_E_TICKET_LENGTH = 40;

/** Letra ou algarismo à cabeça (fecha a porta a fórmulas no Excel), depois só - _ / . */
const E_TICKET = /^[A-Za-z0-9][A-Za-z0-9._/-]*$/;

/** O que está mal no e-Ticket escrito, ou `null` se serve. Vazio serve: apaga. */
export function eTicketProblem(raw: string): string | null {
  const value = raw.trim();
  if (!value) return null;
  if (value.length > MAX_E_TICKET_LENGTH) {
    return `No máximo ${MAX_E_TICKET_LENGTH} caracteres.`;
  }
  if (!E_TICKET.test(value)) {
    return 'Só letras, algarismos e - _ / . — a começar por letra ou algarismo.';
  }
  return null;
}
